### ISSUE TITLE
VMCluster reports app healthy and synced on argocd when the status says otherwise

### ISSUE THREAD

**Issue by nonproto** [VictoriaMetrics/operator#1070]
### Describe the bug

Deploy a VictoriaMetrics Operator and a VMCluster with STS VmSelect.  Make a change to the statefulset declaration that is not allowed and sync.  

Argo shows the app is healthy.  But the VMCluster status shows the
```
status.clusterStatus: failed to create or update vmcluster cannot perform update on sts
```

### To Reproduce

Don't have a config available 

### Version

VMOperator `v0.33.4`
App Version from chart is `v0.46.4`

**Comment by f41gh7** [comment]
Hello, could you please check actual operator logs for the full error text?

**Comment by nonproto** [comment]
here is the sample `vmselect` portion under the `vmcluster`
```
vmselect:
    replicaCount: 1
    revisionHistoryLimitCount: 5
    cacheMountPath: "/select-cache"
    storage:
      volumeClaimTemplate:
        spec:
          storageClassName: my-storage-class
          resources:
            requests:
              storage: 1Gi
```

I deploy the cluster, everything is up and healthy.  I change the revisionHistoryLimitCount from 5 -> 3

ArgoCD shows the VictoriaMetrics App is healthy and synced. 

`vmcluster` is stuck in an expanding status and has the error 
```
Normal   ReconcileEvent      <unknown>  victoria-metrics-operator  starting object update                                                                               
  Warning  ReconcilationError  <unknown>  victoria-metrics-operator  failed create or update vmcluster: cannot perform update on sts: vmselect-victoria-metrics, err: Stat
efulSet.apps "vmselect-victoria-metrics" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy', 'persist
entVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden
```

Operators log show

```
victoria-metrics-controller-vm-7b54fd596b-66jn4 vm {"level":"error","ts":"2024-08-16T12:36:39Z","logger":"manager","msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"victoria-metrics","namespace":"monitoring"},"namespace":"monitoring","name":"victoria-metrics","reconcileID":"14c5cb26-246f-401d-b6ac-730da082cc18","error":"failed create or update vmcluster: cannot perform update on sts: vmselect-victoria-metrics, err: StatefulSet.apps \"vmselect-victoria-metrics\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222"}
```

**Comment by nonproto** [comment]
Yea that's what I did to get by it. I was more concerned that the operator didn't report to argocd that it was unhealthy

**Comment by f41gh7** [comment]
Issue with revisionHistoryLimitCount was  be fixed at https://github.com/VictoriaMetrics/operator/releases/tag/v0.47.0 release


 For argo-cd health checks - it's better to use custom health check lua script for all VM CRD components:

```yaml
data:
  resource.customizations: |
    operator.victoriametrics.com/*e:
      health.lua: |
        hs = {}
        hs.status = "Progressing"
        hs.message = "Waiting for reconcile"
        if obj.status ~= nil then
          if obj.status.status ~= nil then
               if obj.status.status == "operational"
                   hs.status = "Healthy"
                   hs.message = ""
                   return hs
               end
               if obj.status.status == "expanding"
                   return hs
               end
               if obj.status.status == "failed"
                  hs.status = "Degraded"
                   if obj.status.reason ~= nil
                        hs.message = obj.status.reason
                   end
                   if obj.status.lastSyncError ~= nil
                        hs.message = obj.status.lastSyncError
                   end
                  return hs
               end
          end
        end
        return hs
```

**Comment by f41gh7** [comment]
Could  you please share statefulset definition for vmalertmanager/vmstorage?

 Most probably, an issue related to the kubernetes version and default values for statefulset.spec.

**Comment by f41gh7** [comment]
Looks like recent change to the revesionHistoryCount brings more harm than good. Especially when kubernetes doesn't even use this field.

**Comment by f41gh7** [comment]
Issue exists for kubernetes version < 1.27. 

We're going to create patch relase today

**Comment by f41gh7** [comment]
Issue must be fixed at v0.47.3 release

**Comment by passie** [comment]
Maybe I missed it. When will the operator be released together with the helm chart?
latest update on the operator I could find is 0.47.2 in the changelog
